plugins/Unity/Launcher/xdgwatcher.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2019 UBports project.
+ * Author(s): Marius Gripsgard <marius@ubports.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdgwatcher.h"
+
+#include <QDebug>
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QStandardPaths>
+#include <QTextStream>
+
+XdgWatcher::XdgWatcher(QObject* parent)
+    : QObject(parent),
+      m_watcher(new QFileSystemWatcher(this))
+{
+    connect(m_watcher, &QFileSystemWatcher::directoryChanged, this, &XdgWatcher::onDirectoryChanged);
+    connect(m_watcher, &QFileSystemWatcher::fileChanged, this, &XdgWatcher::onFileChanged);
+
+    const auto paths = QStandardPaths::standardLocations(QStandardPaths::ApplicationsLocation);
+    for (const auto &path: paths) {
+        const auto qdir = QDir(path);
+        if (!qdir.exists()) {
+            continue;
+        }
+
+        // Add the path itself to watch for newly added apps
+        m_watcher->addPath(path);
+
+        // Add watcher for eatch app to watch for changes
+        const auto files = qdir.entryInfoList(QDir::Files);
+        for (const auto &file: files) {
+            if (file.suffix() == "desktop") {
+                m_watcher->addPath(file.absoluteFilePath());
+            }
+        }
+    }
+}
+
+// "Ubuntu style" appID is filename without versionNumber after last "_"
+const QString XdgWatcher::stripAppIdVersion(const QString rawAppID) {
+    auto appIdComponents = rawAppID.split("_");
+    appIdComponents.removeLast();
```

Comment:
I think we should only do this, if the `appIdComponents.length() == 3`, otherwise it is not a valid versioned app ID for the purposes here. This would also solve the concern above that we might strip `kde_kate` improperly. It would also simplify the code a fair bit, as we could then avoid special-casing below, and above, only passing the result of `toAppId()` around.

Reply:
Yeah that make sense, will add that :)

Reply:
This has still not been updated?

Reply:
Thinking about this, we may still improperly strip for example `org_kde_kate`, so we would still need to open the file to check